File: TestMergeSort.f

      PROGRAM TMS                              ! main program for test of merge sort

      PARAMETER (M=99999999)

      REAL, ALLOCATABLE :: A(:)
      INTEGER, ALLOCATABLE :: IA(:), IB(:)

      ALLOCATE(A(M))
      ALLOCATE(IA(M))
      ALLOCATE(IB(M))

      DO I=1,M                                 ! initialize array to be sorted
       IF (MOD(I,2).EQ.0) THEN
        A(I)=(M-I)/3.0
       ELSE
        A(I)=(2*I-M)/7.0
       ENDIF
      ENDDO

      CALL CPU_Time(RTIME0)

      CALL MSORT(M,A,IA,IB)                    ! sort indexes of IA array
                                               ! so that A(IA(I)) is sorted
      CALL CPU_Time(RTIME1)
      WRITE(*,710) RTIME1-RTIME0
  710 FORMAT('Time elapsed:',F7.2,'  [sec]')

      write(*,*) A(IA(1)), A(IA(M))

      DEALLOCATE(IB)
      DEALLOCATE(IA)
      DEALLOCATE(A)

      END

!*****************************************************

      SUBROUTINE MSORT(N,A,IB,IC)        !   1  2  3  4  5  6  7  8  9 10 11 12 13 14 15  ...
      REAL A(*)                          ! |--|--|--|--|--|--|--|--|--|--|--|--|--|--|--| ...
      INTEGER IB(*), IC(*)               !  LL          LR          LE                    ... example for K=4
                                         !        SA1         SA2
      DO 1 I=1,N
    1 IB(I)=I                            ! set initial indexes

      K=1                                ! K = subarray (SA) size
    2 IF (K.GT.N) RETURN                 ! outer cycle for increasing SA size
      K2=K+K

      LL=1                               ! LL = left index of the first SA
    3 LR=LL+K; IF (LR.GT.N)   GOTO 4     ! LR = right index of SA; inner cycle
      LE=LR+K; IF (LE.GT.N+1) LE=N+1     ! LE = end of second SA
      M=LL; I=LL; J=LR                   ! indexes for merging SA

      DO WHILE (I.LT.LR.AND.J.LT.LE)     ! merge the first and the second SA into IC
       IF (A(IB(I)).LE.A(IB(J))) THEN
        IC(M)=IB(I); I=I+1               ! move from the first SA
       ELSE
        IC(M)=IB(J); J=J+1               ! move from the second SA
       ENDIF
       M=M+1
      ENDDO

      DO WHILE (I.LT.LR)
       IC(M)=IB(I); I=I+1; M=M+1         ! move the rest of the first SA
      ENDDO

      DO WHILE (J.LT.LE)
       IC(M)=IB(J); J=J+1; M=M+1         ! move the rest of the second SA
      ENDDO

      DO M=LL,LE-1
       IB(M)=IC(M)                       ! move merged part into IB, merging finished
      ENDDO

      LL=LL+K2                           ! set left index of new SA
      GOTO 3                             ! end of inner cycle, go to the next two SA

    4 K=K2                               ! increase SA size two times
      GOTO 2                             ! end of outer cycle

      END

File: TestMergeSort.c

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 99999999

float A[N];
int   B[N],C[N];


void mergesort (long n, float a[], int ib[], int ic[])    //   1  2  3  4  5  6  7  8  9 10 11 12 13 14 15  ...
{int ll, lr, le;                                          // |--|--|--|--|--|--|--|--|--|--|--|--|--|--|--| ...
 int i,j,k,m;                                             //  ll          lr          le                    ... example for k=4
                                                          //        SA1         SA2

 for (i=0; i<n; i++) ib[i]=i;                             // set initial index array
 for (k=1; k < n; k=k+k )                                 // outer cycle for increasing subarray (SA) size k
  {for (ll=0; ll+k < n; ll += k*2 )                       // ll = left index of the first SA
   {lr = ll + k;                                          // lr = right index of SA; inner cycle
    le = lr + k;                                          // le = end of second SA
    if (le > n) le = n;
    m = ll; i = ll; j = lr;
    while (i < lr && j < le)                              // merge the first and the second SA into ic
     {if   (a[ib[i]] <= a[ib[j]]) {ic[m] = ib[i]; i++;}   // move from the first SA
      else                        {ic[m] = ib[j]; j++;}   // move from the second SA
      m++;}
    while (i < lr) {ic[m]=ib[i]; i++; m++;}               // move the rest of the first SA
    while (j < le) {ic[m]=ib[j]; j++; m++;}               // move the rest of the second SA
    for (m=ll; m < le; m++) ib[m] = ic[m];                // move merged part into ib, merging finished
   }
  }
}


int isSorted(float A[], int B[])
{float prev = A[B[0]];                                    // check if the array is sorted
 for (int i = 1; i < N; i++)
  {if (prev > A[B[i]]) {printf("Iterative MergeSort Fails!!"); return 0;}
   prev = A[B[i]];}
 return 1;
}


int main()
{clock_t start, end;
 double cpu_time_used;

 // generate input array
 for (int i = 0; i < N; i++)
  if (i % 3 ==0) A[i] = (N-i)/3.0; else A[i]=(2*i-N)/7.0;

 start = clock();

 // sort array A[0..N-1] using temporary arrays B and C
 mergesort(N, A, B, C);

 end = clock();
 cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;

 // check if array A is sorted via index array B
 if (isSorted(A,B))
  {printf("%f %f\n", A[B[0]],A[B[N-1]]);
   printf("Elapsed time: %f sec\n",cpu_time_used);}

    return 0;
}



